components/exercises/chapter-4-oscar/AList/CityList.tsx

```diff
@@ -0,0 +1,133 @@
+import React, { useState, ReactNode } from 'react';
+import { arrayToList } from './arrayToList';
+import { listToArray } from './listToArray';
+import { prepend } from './prepend';
+import { getFood } from './getFood';
+
+export type List = {
+  value: string | number;
+  rest: List | null;
+};
+
+export const initialItems = arrayToList(['Paris', 'London', 'Barcelona']);
+
+/**
+ * Represents a list item.
+ * @param item - The item to be rendered.
+ * @returns The rendered list item.
+ */
+export function ListItem({ item }: { item: List }): JSX.Element {
+  return (
+    <li>
+      <p className='text-gray-800'>{item.value}</p>
+      {item.rest === null ? (
+        <div className='mb-2'>
+          <div className='cursor-pointer rounded-lg bg-red-600 p-3 shadow'>
+            <p className='font-semibold'>No more cities to visit</p>
+          </div>
+        </div>
+      ) : (
+        <details className='mb-2'>
+          <summary className='cursor-pointer rounded-lg bg-yellow-600 p-3 shadow'>
+            <span className='font-semibold'>
+              {item.rest && `${item.rest.value} and the rest of the cities`}
+            </span>
+          </summary>
+          <div className='bg-white p-4'>
+            {item.rest ? <ListItem item={item.rest} /> : null}
+          </div>
+        </details>
+      )}
+    </li>
+  );
+}
+
+/**
+ * CityList component displays a list of cities and their corresponding countries.
+ * It allows adding new cities to the list randomly.
+ */
+export default function CityList() {
+  const [items, setItems] = useState(initialItems);
+
+  function handleClick(): void {
+    /**
+     * Generates a random city from a predefined list of cities.
+     * @returns {string} The randomly selected city.
+     */
+    function randomCity(): string {
+      const cities = ['Porto', 'Milan', 'Geneva', 'Budapest'];
+      const randomIndex = Math.floor(Math.random() * cities.length);
+      return cities[randomIndex];
+    }
```

Comment:
Can we move this outside the `handleClick`? Or even outside the component, if it doesn't need any binding defined in the component to be in scope.

In general, avoid defining functions inside of functions if possible, it is harder to read

Reply:
I have moved the function to a new file including its test